### PR TITLE
continuoustest: Set both tenant-id and authentication

### LINKED ISSUE #6091: Continuous Test: set both basic-auth and tenant_id
### Discussed in https://github.com/grafana/mimir/discussions/6066

<div type='discussions-op-text'>

<sup>Originally posted by **oujonny** September 19, 2023</sup>
We recently enabled basic auth on the gateway, then the continuous test does not work anymore because we forgot about the auth part there. I was able to configure it with basic_auth, but why can't I then specify a tenant? I would like to use basic auth and the test metrics for a specific tenant (not "anonymous")

see code here: https://github.com/grafana/mimir/blob/c13a9aefd598973c0f66785d876021369bcfaa3d/pkg/continuoustest/client.go#L102C2-L102C2 </div>

Continuous test doesn't support the behaviour of the nginx gateway in the mimir-distributed helm chart. 

Currently, it validates basic auth separately from the the x-scope-orgid HTTP header. 

#### Proposal 1

One solution is to allow setting both in continuous test. I think this is the preferred option

#### Proposal 2

Another is to use the basic auth user as the value of x-scope-org in the nginx config. This will be a breaking change though.

### PR DESCRIPTION
**Pull request by Itaykal**
#### What this PR does
Allows setting tenant-id when writing data using basic-auth/bearer-token in order to test metrics for specific tenants

#### Which issue(s) this PR fixes or relates to

Fixes #6091

#### Checklist

- [x] Tests updated.
- [ ] Documentation added.
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [ ] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

**Review by 56quarters — CHANGES_REQUESTED** [review]
Can you explain why you'd want to use both tenant ID and auth at the same time? This doesn't make sense to me.

In Mimir, you'd only set the tenant ID via header.

With Grafana Enterprise Metrics you'd set basic auth and the authentication layer would resolve that to a tenant ID.

Adding the ability to set both of them seems like it will introduce confusion.

**Comment by Itaykal** [comment]
> Can you explain why you'd want to use both tenant ID and auth at the same time? This doesn't make sense to me.

The official Helm chart supports configuring a gateway that respects this behavior (basic-auth + custom tenant-id), it would be quite frustrating if other tools in the ecosystem such as continuous test won't support this.

And even if you don't use the helm chart specified here, It's not a stretch to assume some sort of authentication is required to reach the cluster.

> In Mimir, you'd only set the tenant ID via header.

In some environments this can't be the only security measure and basic auth or any other authentication solution could be used (without coupling them to tenants as done in GEM).

> Adding the ability to set both of them seems like it will introduce confusion.

I can see why you'd say that, but I think re-creating the same behavior on non GEM instances will be a more complicated solution prone to breaking changes on some environments. 

I understand your point of this not being a common use-case but I've encountered it, as well as the issue's creator, so I guess this is a thing people do

**Review by 56quarters — APPROVED** [review]
Sorry for blocking this, your explanation makes sense. Can you add an entry to the changelog that mentions that tenant ID and basic/bearer auth is now allowed to be combined?

**Comment by Itaykal** [comment]
> Sorry for blocking this, your explanation makes sense. Can you add an entry to the changelog that mentions that tenant ID and basic/bearer auth is now allowed to be combined?

Done

File: CHANGELOG.md [hunk 1]
```diff
@@ -81,6 +81,7 @@
 * [CHANGE] `mimir-continuous-test` has been deprecated and replaced by a Mimir module that can be run as a target from the `mimir` binary using `mimir -target=continuous-test`. #7753
 * [CHANGE] `-server.metrics-port` flag is no longer available for use in the module run of mimir-continuous-test, including the grafana/mimir-continuous-test Docker image which uses the new module. Configuring this port is still possible in the binary, which is deprecated. #7747
 * [BUGFIX] Set `User-Agent` header for all requests sent from the testing client. #7607
+* [CHANGE] Allowed authenticatication to Mimir using both Tenant ID and basic/bearer auth #6091.
```
Comment (56quarters) [inline CHANGELOG.md:84]:
```suggestion
* [CHANGE] Allowed authenticatication to Mimir using both Tenant ID and basic/bearer auth #7619.
```

**Comment by kressnick25** [comment]
Continuous test client appears to still only set basic-auth OR tenant ID if both are defined - per
https://github.com/grafana/mimir/blob/main/pkg/continuoustest/client.go#L268-L274
```go
if rt.bearerToken != "" {
    req.Header.Set("Authorization", "Bearer "+rt.bearerToken)
} else if rt.basicAuthUser != "" && rt.basicAuthPassword != "" {
    req.SetBasicAuth(rt.basicAuthUser, rt.basicAuthPassword)
} else {
    req.Header.Set("X-Scope-OrgID", rt.tenantID)
}
```

When passing both username/password AND tenantId to the latest pre-release binary, I am receiving
` status_code=401 err="server returned HTTP status 401 Unauthorized and body \"no org id\\n\"` response in the `mimir-continous-test` output

**Comment by Itaykal** [comment]
@kressnick25 Could you create an issue about this? I'll try to find time to work on it in the following days but I'm not sure I could

**Pull request merged**